Dev

PetrNovakPN opened the pull request.

PetrNovakPN committed:
Added project in progress

PetrNovakPN committed:
rm init weather, adjusted constructors

removed initial weather part, adjusted constructors

PetrNovakPN committed:
Models adjustments

price double => decimal, removed join table, connected db

PetrNovakPN committed:
Project restructure

Restructured project into folders for better readability and accuracy

PetrNovakPN committed:
Small code clarity change

var => types

Steffik369 commented on Wise_Owl_Library/Models/Book.cs:
```diff
@@ -0,0 +1,24 @@
+        public int Stock { get; set; }
+
+        [Required]
+        public List<Author> Authors { get; set; } = [];
```
Možná by byla dobrá validace aby měl alespoň jednoho autora

PetrNovakPN replied:
Jo to mi tam uteklo jak jsem odebíral BookAuthor

Steffik369 commented on Wise_Owl_Library/Models/PriceChange.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+        public int Id { get; set; }
+        public int BookId { get; set; }
+        public required Book Book { get; set; }
```
Buď bych tu nechal jen book a nebo jen to id.

Steffik369 commented on Wise_Owl_Library/Models/PriceChange.cs:
```diff
@@ -0,0 +1,19 @@
+        [Column(TypeName = "decimal(18,2)")]
+        public decimal NewPrice { get; set; }
+
+        public DateTime ChangeDate { get; set; }
```
DateTimeOffset je lepší

Steffik369 commented on Wise_Owl_Library/Models/Author.cs:
```diff
@@ -0,0 +1,16 @@
+        public required string Name { get; set; }
+
+        [Required]
+        public List<Book> Books { get; set; } = [];
```
Tohle je tu možná cyklická vazba, bude stačit udělat get metodu v servise pokud to bude potřeba

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,206 @@
+            {
+                Book? book = await context.Books
+                    .Include(b => b.Authors)
+                    .FirstOrDefaultAsync(b => b.Id == id);
```
Find je rychlejší

PetrNovakPN committed:
Refactor

Ensure services operate on Book model rather than DTOs

PetrNovakPN committed:
feat: add middleware and exception handlers for error handling

- Implemented custom middleware for centralized error logging.
- Added exception handling for DbUpdateException, TimeoutException, etc.

PetrNovakPN committed:
test: add unit tests for BooksController, ErrorLoggingMiddleware, and PriceChangeController

Added unit tests for the following components:
- BooksController*: Verifies endpoint responses for book retrieval, creation, updating, and deletion, ensuring correct handling of various inputs and states.
- **ErrorLoggingMiddleware**: Tests logging behavior and error handling, confirming that the middleware correctly logs exceptions and sets appropriate response codes and formats.
- **PriceChangeController**: Ensures accurate price change handling, validating that the controller processes price updates and applies business rules as expected.

Steffik369 commented on Wise_Owl_Library/Controllers/BooksController.cs:
```diff
@@ -0,0 +1,100 @@
+            {
+                return Ok(new List<BookDto>());
+            }
+            return Ok(books.Select(book => new BookDto(book)).ToList());
```
Tenhle if je tu zbytečný, půjde to napsat celé do jednoho řádku přes ten select

Steffik369 commented on Wise_Owl_Library/Controllers/BooksController.cs:
```diff
@@ -0,0 +1,100 @@
+        [HttpGet("{id}")]
+        public async Task<ActionResult<BookDto>> GetBook(int id)
+        {
+            Book book = await bookService.GetBookAsync(id) ?? throw new KeyNotFoundException($"Book with ID {id} not found.");
```
Tohle asi bude lepší spíš vrátit jako result ne jako chybu. I to že to nenašel není chyba ale jen stav. Chyba by to byla pokud by to tam nutně muselo být a nemohl by nastat stav že to tam není.

Steffik369 replied:
Napsal jsem to jen sem ale asi bych to upravil všude

Steffik369 commented on Wise_Owl_Library/Controllers/BooksController.cs:
```diff
@@ -0,0 +1,100 @@
+                Price = dto.Price,
+                Stock = dto.Stock,
+                Authors = dto.Authors.Select(a => new Author { Name = a.Name }).ToList()
+            }).ToList();
```
Tohle bych taky přepsal do extension metody. Ať je to na jednom místě a tady to bude čitelnější

Steffik369 commented on Wise_Owl_Library/Controllers/BooksController.cs:
```diff
@@ -0,0 +1,100 @@
+            if (id != updateBookDto.Id)
+            {
+                throw new ArgumentException("The provided ID does not match the book ID.");
+            }
```
Zdá se že to tam máš zbytečně dvakrát to id. Stačí to v modelu 😉

Steffik369 commented on Wise_Owl_Library/Controllers/BooksController.cs:
```diff
@@ -0,0 +1,100 @@
+                Authors = updateBookDto.Authors.Select(a => new Author { Name = a.Name }).ToList()
+            };
+
+            bool result = await bookService.UpdateBookAsync(id, updatedBook);
```
Špatně pojmenovaný. Result je něco jiného, když je to bool tak spíš něco jako wasUpdated atd -> viz naming konvence boolu

Steffik369 commented on Wise_Owl_Library/Controllers/BooksController.cs:
```diff
@@ -0,0 +1,100 @@
+            };
+
+            bool result = await bookService.UpdateBookAsync(id, updatedBook);
+            if (!result)
```
Tady pak není jasné kvůli tomu pojmenování jestli checkujes async result a může to mást

Steffik369 commented on Wise_Owl_Library/Controllers/PriceChangeController.cs:
```diff
@@ -0,0 +1,31 @@
+            List<PriceChangeDto> priceChangeDtos;
+
+           
+            priceChangeDtos = await priceChangeService.GetPriceChangesAsync();
```
Tohle klidně zdrcni na jeden řádek

Steffik369 commented on Wise_Owl_Library/Controllers/PriceChangeController.cs:
```diff
@@ -0,0 +1,31 @@
+
+            if (priceChangeDtos.Count == 0)
+            {
+                throw new KeyNotFoundException("No price changes found.");
```
Zase tohle není chyba ale validní výsledek

Steffik369 commented on Wise_Owl_Library/Data/Dto/BookDto.cs:
```diff
@@ -0,0 +1,13 @@
+        public decimal Price { get; set; } = book.Price;
+        public int Stock { get; set; } = book.Stock;
+        public List<string> Authors { get; set; } = book.Authors.Select(a => a.Name).ToList();
+    }
```
Ten konstruktor nech prázdný a udělej extension metodu. Je to univerzálnější

Steffik369 replied:
Jen tam pak doplň místo toho required kde to bude potřeba

Steffik369 commented on Wise_Owl_Library/Interfaces/IBookService.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public interface IBookService
+    {
+        Task<IEnumerable<Book>> GetBooksAsync(string? title, int? stock);
```
Pokud to uživatel nemusí vyplňovat klidně bych tam dal default null, bude to zřejmější že nemusí

Steffik369 commented on Wise_Owl_Library/Interfaces/IBookService.cs:
```diff
@@ -0,0 +1,15 @@
+    public interface IBookService
+    {
+        Task<IEnumerable<Book>> GetBooksAsync(string? title, int? stock);
+        Task<Book?> GetBookAsync(int id);
```
Tady to je krásně vidět, máš tam book? To znamená že je validní že žádná nepřijde. Tudíž exception když nenajde nedává smysl. To by tu musel být jen book. Ale jak jsem psal mě to přijde validní že to tam není takže prostě vrátí 404

Steffik369 commented on Wise_Owl_Library/Interfaces/IBookService.cs:
```diff
@@ -0,0 +1,15 @@
+        Task<IEnumerable<Book>> GetBooksAsync(string? title, int? stock);
+        Task<Book?> GetBookAsync(int id);
+        Task<IEnumerable<Book>> CreateBooksAsync(List<Book> books);
+        Task<bool> UpdateBookAsync(int id, Book updatedBook);
```
Tady je ta duplikace jak jsem psal

Steffik369 commented on Wise_Owl_Library/Interfaces/IPriceChangeService.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IPriceChangeService
+    {
+        Task<List<PriceChangeDto>> GetPriceChangesAsync();
```
Tady by nemělo být dto

Steffik369 commented on Wise_Owl_Library/Middleware/ErrorLoggingMiddleware.cs:
```diff
@@ -0,0 +1,47 @@
+                var (statusCode, title, detail) = ex switch
+                {
+                    ArgumentException argEx => (StatusCodes.Status400BadRequest, "Invalid parameters", argEx.Message),
+                    KeyNotFoundException keyEx => (StatusCodes.Status404NotFound, "Resource not found", keyEx.Message),
```
Aaaa chápu. Tím pádem může být. Jen pozor na to že exceptions jsou drahý, takže validní stavy je lepší řešit ne přes ně ale logikou

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,179 @@
+
+                if (stock.HasValue)
+                {
+                    query = query.Where(b => b.Stock == stock.Value);
```
Tohle mi moc nedává smysl businessove, proč by někdo chtěl hledat přesný počet a ne něco jako víc jak 5 třeba?

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,179 @@
+            {
+                logger.LogError(ex, "Error loading books.");
+                throw;
+            }
```
Ty exceptiony bych řešil až na úrovni kontroleru

Steffik369 replied:
A na úrovni db layeru - repository pattern

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,179 @@
+                {
+                    if (await BookExistsAsync(book.Title, book.Authors.Select(a => a.Name).ToList()))
+                    {
+                        throw new InvalidOperationException($"The book '{book.Title}' already exists.");
```
Je tohle chyba nebo by to měl prostě jen vynechat / updatnout?

Steffik369 replied:
Záleží na business logice k čemu to slouží a co uživatel čeká

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,179 @@
+            if (id != updatedBook.Id)
+            {
+                throw new ArgumentException("ID mismatch.");
+            }
```
Někde už jsem to viděl, jen trochu jinak-> duplikát + zbytečně je to tu dvakrát, stačí to id v modelu

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,179 @@
+            {
+                Book? book = await context.Books
+                    .Include(b => b.Authors)
+                    .FirstOrDefaultAsync(b => b.Id == id);
```
Find asi lepší

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,179 @@
+            }
+        }
+
+        public async Task<bool> UpdateBookAsync(int id, Book updatedBook)
```
Většinou se to dělá tak že se vrací id v případě úspěchu, aby s tím uživatel mohl pracovat dal. Nebo taky uodatnuta entita

Steffik369 replied:
Ale klidně může být i takhle

Steffik369 commented on Wise_Owl_Library/Services/PriceChangeService.cs:
```diff
@@ -0,0 +1,32 @@
+                OldPrice = pc.OldPrice,
+                NewPrice = pc.NewPrice,
+                ChangeDate = pc.ChangeDate
+            }).ToList();
```
Tohle až vejs jak jsem říkal, na level kontroleru

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,179 @@
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(ex, "Error updating book.");
```
Když už tak aspoň doplnit id nebo něco ať se to dobře debuguje

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,179 @@
+
+        private async Task<bool> BookExistsAsync(string title, List<string> authorNames)
+        {
+            return await context.Books.AnyAsync(b => b.Title == title && b.Authors.Any(a => authorNames.Contains(a.Name)));
```
Hledat to takhle je asi trochu divný. Ideálně by bylo dobré podle ISBN

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,179 @@
+            context.PriceChanges.Add(priceChange);
+        }
+
+        private static void UpdateBookDetails(Book book, Book updatedBook)
```
Tohle není dobré řešení, nemělo by se to používat takto, protože to je nečitelný co to s tím vevnitř udělá. Mělo by to aspoň vracet book ne void.

Steffik369 reviewed (commented):
Tests review

PetrNovakPN committed:
feat(pricechange, book): change Changedate to datetimeoffset, add author validation, and refactor code

-Updated the PriceChange model to use datetimeoffset for the Changedate property to improve timezone handling.
-Added validation to ensure that each book has at least one author to maintain data integrity.
-Also refactored some code

PetrNovakPN committed:
refactor(pricechange): separate service and controller logic with distinct model handling

-Aligned the PriceChange service to work exclusively with domain models and the controller with database models.
-Updated the interface between them to ensure seamless integration and adjusted the dependent code for compatibility.

PetrNovakPN committed:
feat(PriceChangeService, validation, refactor): update ChangeDate type, add author validation, and refactor code

- Changed ChangeDate from DateTime to DateTimeOffset in PriceChange model to support timezone information.
- Added validation to require at least one author for each book to maintain data integrity.
- Refactored code in PriceChangeService, PriceChangeController, and BookController for improved readability and maintainability.
- Added extension methods for Dto mappings and updated unit tests for new service logic.

PetrNovakPN committed:
refactor: update service methods to return modified objects and improve error handling

- Updated `UpdateBookAsync` to return the modified `Book` object instead of a boolean, providing more information to the calling functions and enhancing error feedback.
- Refactored `CreateBooksAsync` to return a list of successfully created books and handle duplicate checks individually for each book, with improved error handling and logging.
- Removed the `UpdateBookDetails` helper method, moving book property updates directly into `UpdateBookAsync`.
- Standardized error messages and improved logging for clearer debugging, including specific identifiers for problematic book operations.

PetrNovakPN committed:
BREAKING CHANGE: Adopting the Repository Pattern changes the structure of data access and modifies how repositories are used across the application.

PetrNovakPN committed:
Refactor BooksController and update service methods

-Refactored the `BooksController` to improve handling of `Book` and `BookDto` objects.
-Updated `GetBook`, `PostBooks`, `PutBook`, and `DeleteBook` methods for better data conversion and response handling.
-Added new extension methods for DTO to model conversions.
-Modified `IBookRepository` and `BookService` to support batch operations and return meaningful results.
-Removed the `Books` property from the `Author` model and fixed formatting in the `Book` model.

PetrNovakPN committed:
feat(books): implement repository methods and improve BookDto mapping

C-ompleted implementations for repository methods such as `DeleteBookAsync`, `GetBookAsync`, and `UpdateBookAsync`.
-Enhanced `GetBooksAsync` to handle filters by title and stock.
-Fixed BookDto transformation in the controller and optimized LINQ queries.

PetrNovakPN committed:
feat: added entity models

PetrNovakPN committed:
Added azure sqldb

PetrNovakPN committed:
refactor(BookSqlRepository): So it uses entity models

PetrNovakPN committed:
feat: enhance API responses and migrate database

- Improved `BooksController.cs` responses with detailed messages (e.g., "No books found") and adjusted update/delete responses.
- Enhanced `ErrorLoggingMiddleware.cs` to handle `KeyNotFoundException` with HTTP 404 status.
- Removed old migration files and updated `WiseOwlLibraryDbContextModelSnapshot.cs` with new entities.
- Added `Id` to `UpdateBookDto.cs` and initialized collections in `AuthorEntity.cs` and `BookEntity.cs`.
- Registered `PriceChangeInterceptor` in `Program.cs` and updated connection string to `DefaultConnection` for temporary local db.
- Enhanced `BookSqlRepository.cs` and `PriceChangeSqlRepository.cs` methods for managing authors and price changes.
- Implemented `DetectPriceChanges` in `PriceChangeInterceptor.cs` to track price updates.
- Added `PriceChangeExtension.cs` for mapping `PriceChange` and `PriceChangeEntity`.
- Created migrations for table updates and model changes in `20241123130810_Initial` and `20241124000541_PriceChangeEntity`.

Steffik369 commented on Wise_Owl_Library/Extensions/BookEntityExtension.cs:
```diff
@@ -0,0 +1,17 @@
+            Title = book.Title,
+            Price = book.Price,
+            Stock = book.Stock,
+            Authors = book.Authors.Select(a => new AuthorEntity { Id = a.Id, Name = a.Name }).ToList()
```
Tady můžeš použít taky ten mapper co máš v AuthorExtension. Stejně tak v ostatních mapperech

Steffik369 commented on Wise_Owl_Library/Extensions/BookExtension.cs:
```diff
@@ -0,0 +1,27 @@
+            Authors = bookDto.Authors.Select(a => new Author { Id = a.Id, Name = a.Name }).ToList()
+        };
+
+        public static Book EntityToBook(this BookEntity bookEntity) => new()
```
Ten název ti tu nesedí protože tohle by mělo být pravděpodobně v jiné třídě. Samostatné extension metodě pro BookEntity

Steffik369 replied:
Asi bych to rozdělil a udělal 2 mappery jeden pro dto layer a druhý pro db layer

Steffik369 commented on Wise_Owl_Library/Models/Author.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public int Id { get; set; }
+
+        [Required(ErrorMessage = "Name is required.")]
+        [StringLength(50, ErrorMessage = "Name length can't be more than 50 characters.")]
```
Podle mě tahle validace nepokryje případ kdy tam bude empty nebo whitespace - nevím jestli to přes artributy jde. Já na to používám Custom mappery kde mám pak vše na jednom místě

Steffik369 commented on Wise_Owl_Library/Repositories/BookSqlRepository.cs:
```diff
@@ -0,0 +1,121 @@
+        public async Task<bool> DeleteBookAsync(int id)
+        {
+            BookEntity? book = await wiseOwlLibraryDbContext.Books.FindAsync(id);
+            if (book == null)
+            {
+                throw new KeyNotFoundException($"Book with Id {id} not found.");
+            }
```
To že tam není nemusí nutně znamenat chybu, záleží jestli to máš stateless nebo stateful

Steffik369 commented on Wise_Owl_Library/Repositories/BookSqlRepository.cs:
```diff
@@ -0,0 +1,121 @@
+            BookEntity? book = await wiseOwlLibraryDbContext.Books
+                .Include(b => b.Authors)
+                .FirstOrDefaultAsync(b => b.Id == id);
+            if (book == null)
+            {
+                throw new KeyNotFoundException($"Book with Id {id} not found.");
+            }
```
Tohle taky nemusí být nutně chyba. Často se ta metoda prepouziva v exist checku, kde tonoak nechceš handlovat přes exceptiony. Až ten kdo to odebírá by si měl říct jestli to je chyba

Steffik369 commented on Wise_Owl_Library/Repositories/BookSqlRepository.cs:
```diff
@@ -0,0 +1,121 @@
+            return false;
+        }
+
+        public async Task<bool> BookExistsByIdAsync(int id)
+        {
+            if (await wiseOwlLibraryDbContext.Books.AnyAsync(b => b.Id == id))
+            {
+                return true;
+            }
+            return false;
+        }
```
Tady bych právě prepouzil už hotové implementace na get

Steffik369 commented on Wise_Owl_Library/Repositories/PriceChangeSqlRepository.cs:
```diff
@@ -0,0 +1,40 @@
+{
+    public interface IPriceChangeRepository
+    {
+        /// <summary>
+        /// cokoliv
+        /// </summary>
+        /// <returns></returns>
```
Tohle vyhod, čitelný kód nepotřebuje komentáře. + Tohle je v aktuálním stavu k ničemu

Steffik369 commented on Wise_Owl_Library/Repositories/WiseOwlLibraryDbContext.cs:
```diff
@@ -0,0 +1,31 @@
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            modelBuilder.Entity<BookEntity>()
+                .HasMany(b => b.Authors);
+
+            modelBuilder.Entity<AuthorEntity>()
+                .HasMany(a => a.Books);
+
+            base.OnModelCreating(modelBuilder);
```
Mě se ta vazba many to many zda pořád moc na tenhle případ. Zbytečně složité a náchylné na chyby. Při množství záznamu bude efektivnější naimplementovat to bez toho a řešit to jen v repository jednou "neoptimální" query

Steffik369 replied:
Do produkce bych to asi nedal. Jednodušší kód je lepší kód

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,79 @@
+        public async Task<Book?> GetBookAsync(int id)
+        {
+            Book? book = await bookRepository.GetBookAsync(id);
+            if (book == null)
+            {
+                return null;
+            }
+            return book;
```
Dej sem ternární operátor, bude to hezčí a čitelnější

Steffik369 commented on Wise_Owl_Library/Services/BookService.cs:
```diff
@@ -0,0 +1,79 @@
+                }
+            }
+            List<Book> createdBooks = await bookRepository.AddBooksAsync(booksToCreate);
+            if (createdBooks == null)
+            {
+                return [];
+            }
+
+            return createdBooks;
```
Tady asi taky ternary

Steffik369 replied:
A obecně možná na víc míst

PetrNovakPN committed:
refactor: extension methods and organize them into separate folders

PetrNovakPN committed:
refactor: improve book and author handling

- Added `ToBook` methods in `BookPresentationExtensions` for converting `BookDto` and `UpdateBookDto`.
- Introduced `NoEmptyOrExcessiveSpaces` validation attribute for string validation in `AuthorDto` and `CreateBookDto`.
- Consolidated `UpdateBookDtoExtensions` functionality into `BookPresentationExtensions`.
- Refactored repository methods: `DeleteBookAsync`, `BookExistsAsync`, and `UpdateBookAsync` in `BookSqlRepository`.
- Updated `BookService` to utilize the refactored `BookExistsAsync`.

PetrNovakPN committed:
Refactor data annotations and handling of missing books

Steffik369 commented on Wise_Owl_Library/Controllers/BooksController.cs:
```diff
@@ -19,7 +19,7 @@ public async Task<ActionResult<BookDto[]>> GetBooks(string? title = null, int? s
             if(books.Count == 0)
             {
-                return NoContent();
+                return Ok(new { message = "No books found." });
```
Tohle klidně mohlo být NoContent

Steffik369 commented on Wise_Owl_Library/Repositories/WiseOwlLibraryDbContext.cs:
```diff
@@ -13,10 +14,18 @@ public class WiseOwlLibraryDbContext(DbContextOptions<WiseOwlLibraryDbContext> o
+                .HasMany(b => b.Authors);
+
+            modelBuilder.Entity<AuthorEntity>()
+                .HasMany(a => a.Books);
```
Tohle je zvláštní vazba, teď to tam bude vypadat že to je propojené, ale ve skutečnosti tam v db vazba reálně nebude ne? Nebo si s tím EF poradí? Docela by mě zajímalo jak budou vypadat constrains a schéma. a jak se to zachová např při smazání pouze autora, nebo pouze knihy.
